## Segmenting and Clustering Neighborhoods in Toronto - Project

Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [14]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Libraries imported.


In [2]:
pip install html5lib

Note: you may need to restart the kernel to use updated packages.


In [3]:
pip install lxml

     |████████████████████████████████| 5.5MB 6.1MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


## 1. Download and Explore Dataset

The project uses the Toronto neighborhood data extracted from: https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M,

In [1]:
import pandas as pd

url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
dfs = pd.read_html(url)
dfs[0]

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                          Neighborhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

* The dataframe consists of three columns: PostalCode, Borough, and Neighborhood
* Only processing the cells that have an assigned borough and ignoring cells with a borough that is "Not assigned".
* More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.

In [3]:
toronto_data = dfs[0][dfs[0]['Borough']!='Not assigned']
toronto_data =  toronto_data.reset_index(drop=True)


In [4]:
toronto_data.shape

(103, 3)

In [5]:
toronto_data.describe()

Postal Code     Borough Neighborhood
count          103         103          103
unique         103          10           99
top            M4G  North York    Downsview
freq             1          24            4

In [6]:
toronto_data.head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

As you can see the shape of the final processes data has 103 unique postal codes, 10 unique Boroughs and 99 unique Neigbhorhoods**

*Note: Neighborhood as explained above has instances with shared postal code as in the case of M5A, this is the current state and will be enhanced in the code below*

### 1.1. Converting postal codes and neighborhoods to Latitude and Longitude coordinates


In [7]:
!wget -q -O 'Geospatial_Coordinates (1).csv' http://cocl.us/Geospatial_data
print('Data downloaded!')

Data downloaded!


In [8]:
toronto_lat_long = pd.read_csv('Geospatial_Coordinates (1).csv')
toronto_lat_long.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [9]:
toronto_final = toronto_data.join(toronto_lat_long.set_index('Postal Code'), on='Postal Code')

In [10]:
toronto_final.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [11]:
toronto_final.shape
# toronto_final[toronto_final['Neighborhood']=='Kingsview Village, St. Phillips, Martin Grove Gardens, Richview Gardens']

(103, 5)

In [15]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="ca_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


Visualizing the Toronto coordinates using the Folium map on the overall data for dataframe "toronto_final"

In [16]:
# create map of Canada using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_final['Latitude'], toronto_final['Longitude'], toronto_final['Borough'], toronto_final['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=10,
        popup=label,
        color='white',
        fill=True,
        fill_color='#ea4335',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### 1.2. Toronto boroghs Clustering
* We will identify the boroughs containing the key word "Toronto" and filter a new data frame
* The visualization will performed below for the new dataframe
* Foursquare will be used to explore the neighborhoods around the filtered dataframe
* We will cluster the updated data to identify similar such neighborhoods

In [17]:
toronto_rev = toronto_final[toronto_final['Borough'].str.contains("Toronto")].reset_index(drop=True)
toronto_rev.head()

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

In [52]:
toronto_rev.shape

(39, 5)

Updated map visualization for the new dataframe, includes only the Boroughs containing "Toronto" as the key word

In [18]:
# Updated map for the new data frame
map_toronto_rev = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_rev['Latitude'], toronto_rev['Longitude'], toronto_rev['Borough'], toronto_rev['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=10,
        popup=label,
        color='white',
        fill=True,
        fill_color='#4b63f3',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto_rev)  
    
map_toronto_rev

#### Define Foursquare Credentials and Version

In [19]:
CLIENT_ID = 'UB34IKK4RBRGH01YWR31UD4WQNNUJX0EWP0LB3J2HORJKYP1' # your Foursquare ID
CLIENT_SECRET = 'X4PVN1DAD4JTLZA4IVR50R13VCT4SQRSOUUYLBAQQ5KY10FT' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
radius=500
LIMIT=100

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: UB34IKK4RBRGH01YWR31UD4WQNNUJX0EWP0LB3J2HORJKYP1
CLIENT_SECRET:X4PVN1DAD4JTLZA4IVR50R13VCT4SQRSOUUYLBAQQ5KY10FT


In [20]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items'] # mistake logged here needs to be venues
        
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [21]:
toronto_venues = getNearbyVenues(names=toronto_rev['Neighborhood'],
                                   latitudes=toronto_rev['Latitude'],
                                   longitudes=toronto_rev['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

In [22]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4          Impact Kitchen       43.656369       -79.356980   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4           Restaurant

In [23]:
toronto_venues.shape

(1625, 7)

In [24]:
print('There are {} uniques venue categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 232 uniques venue categories.


In [48]:
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
fixed_columns[fixed_columns.index('Neighborhood')],fixed_columns[fixed_columns.index('Yoga Studio')]=fixed_columns[fixed_columns.index('Yoga Studio')],fixed_columns[fixed_columns.index('Neighborhood')]
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()


Neighborhood  Airport  Airport Food Court  Airport Lounge  \
0  Regent Park, Harbourfront        0                   0               0   
1  Regent Park, Harbourfront        0                   0               0   
2  Regent Park, Harbourfront        0                   0               0   
3  Regent Park, Harbourfront        0                   0               0   
4  Regent Park, Harbourfront        0                   0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0         0            0           0                    0                 0   
1         0            0           0                    0                 0   
2         0            0           0                    0                 0   
3         0            0           0                    0                 0   
4         0            0           0                    0                 0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       1     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   0      0                0         0   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bistro  Board Shop  Boat or Ferry  \
0           0                   0       0           0              0   
1           0                   0       0           0              0   
2           0                   0       0           0              0   
3           0                   0       0           0              0   
4           0                   0       0           0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               0        0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  Butcher  \
0                0         0             0              0         0        0   
1                0         0             0              0         0        0   
2                0         0             0              0         0        0   
3                0         0             0              0         0        0   
4                0         0             0              0         0        0   

   Café  Cajun / Creole Restaurant  Camera Store  Candy Store  \
0     0                          0             0            0   
1     0                          0             0            0   
2     0                          0             0            0   
3     0                          0             0            0   
4     0                          0       

In [49]:
toronto_onehot.shape

(1625, 232)

In [51]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped


Neighborhood   Airport  \
0                                         Berczy Park  0.000000   
1        Brockton, Parkdale Village, Exhibition Place  0.000000   
2   Business reply mail Processing Centre, South C...  0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...  0.071429   
4                                  Central Bay Street  0.000000   
5                                            Christie  0.000000   
6                                Church and Wellesley  0.000000   
7                      Commerce Court, Victoria Hotel  0.000000   
8                                          Davisville  0.000000   
9                                    Davisville North  0.000000   
10                       Dufferin, Dovercourt Village  0.000000   
11             First Canadian Place, Underground city  0.000000   
12    Forest Hill North & West, Forest Hill Road Park  0.000000   
13                           Garden District, Ryerson  0.000000   
14  Harbourfront East, Union Station, Toronto Islands  0.000000   
15                      High Park, The Junction South  0.000000   
16                     India Bazaar, The Beaches West  0.000000   
17          Kensington Market, Chinatown, Grange Park  0.000000   
18                                      Lawrence Park  0.000000   
19                           Little Portugal, Trinity  0.000000   
20                        Moore Park, Summerhill East  0.000000   
21                  North Toronto West, Lawrence Park  0.000000   
22                             Parkdale, Roncesvalles  0.000000   
23        Queen's Park, Ontario Provincial Government  0.000000   
24                          Regent Park, Harbourfront  0.000000   
25                           Richmond, Adelaide, King  0.000000   
26                                           Rosedale  0.000000   
27                                           Roselawn  0.000000   
28                                 Runnymede, Swansea  0.000000   
29                                     St. James Town  0.000000   
30                        St. James Town, Cabbagetown  0.000000   
31                                     Stn A PO Boxes  0.000000   
32                                    Studio District  0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...  0.000000   
34                The Annex, North Midtown, Yorkville  0.000000   
35                                        The Beaches  0.000000   
36                       The Danforth West, Riverdale  0.000000   
37           Toronto Dominion Centre, Design Exchange  0.000000   
38                     University of Toronto, Harbord  0.000000   

    Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0             0.000000        0.000000         0.000000          0.000000   
1             0.000000        0.000000         0.000000          0.000000   
2             0.000000        0.000000         0.000000          0.000000   
3             0.071429        0.071429         0.214286          0.071429   
4             0.000000        0.000000         0.000000          0.000000   
5             0.000000        0.000000         0.000000          0.000000   
6             0.000000        0.000000         0.000000          0.000000   
7             0.000000        0.000000         0.000000          0.000000   
8             0.000000        0.000000         0.000000          0.000000   
9             0.000000        0.000000         0.000000          0.000000   
10            0.000000        0.000000         0.000000          0.000000   
11            0.000000        0.000000         0.000000          0.000000   
12            0.000000        0.000000         0.000000          0.000000   
13            0.000000        0.000000         0.000000          0.000000   
14            0.000000        0.000000         0.000000          0.000000   
15            0.000000        0.000000         0.000000          0.000000   
16            0.000000        0.000000         0.00

#### Let's print each neighborhood along with the top 5 most common venues

In [53]:
num_top_venues = 5
###### sort_values()
for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
            venue  freq
0     Coffee Shop  0.09
1    Cocktail Bar  0.05
2          Bakery  0.03
3  Farmers Market  0.03
4     Cheese Shop  0.03


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.13
1  Breakfast Spot  0.09
2     Coffee Shop  0.09
3      Restaurant  0.04
4         Stadium  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                venue  freq
0  Light Rail Station  0.11
1       Burrito Place  0.06
2          Skate Park  0.06
3          Restaurant  0.06
4      Farmers Market  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                 venue  freq
0      Airport Service  0.21
1              Airport  0.07
2      Harbor / Marina  0.07
3  Rental Car Location  0.07
4          Coffee Shop  0.07


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0

#### Let's put that into a *pandas* dataframe
First, let's write a function to sort the venues in descending order.

In [54]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Creating the dataframe to convert the grouped data

In [55]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head(10)

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   
5                                           Christie         Grocery Store   
6                               Church and Wellesley           Coffee Shop   
7                     Commerce Court, Victoria Hotel           Coffee Shop   
8                                         Davisville          Dessert Shop   
9                                   Davisville North                 Hotel   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0          Cocktail Bar              Pharmacy     Seafood Restaurant   
1        Breakfast Spot           Coffee Shop  Performing Arts Venue   
2            Comic Shop                Garden             Skate Park   
3               Airport         Boat or Ferry                  Plane   
4    Italian Restaurant        Sandwich Place    Japanese Restaurant   
5                  Café                  Park              Nightclub   
6      Sushi Restaurant   Japanese Restaurant             Restaurant   
7            Restaurant                  Café                  Hotel   
8        Sandwich Place      Sushi Restaurant            Coffee Shop   
9  Gym / Fitness Center        Breakfast Spot      Food & Drink Shop   

  5th Most Common Venue 6th Most Common Venue     7th Most Common Venue  \
0                Bakery            Restaurant                  Beer Bar   
1             Pet Store         Burrito Place                Restaurant   
2         Burrito Place        Farmers Market                       Spa   
3           Coffee Shop   Rental Car Location          Sculpture Garden   
4                  Café                   Bar               Salad Place   
5                 Diner           Candy Store        Italian Restaurant   
6               Gay Bar           Men's Store  Mediterranean Restaurant   
7   American Restaurant                   Gym        Seafood Restaurant   
8           Pizza Place                   Gym                      Café   
9        Sandwich Place      Department Store                      Park   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                  Café           Cheese Shop         Farmers Market  
1          Climbing Gym     Convenience Store                    Bar  
2  Fast Food Restaurant               Butcher             Restaurant  
3              Boutique       Harbor / Marina     Airport Food Court  
4       Bubble Tea Shop       Thai Restaurant           Burger Joint  
5            Baby Store            Restaurant            Coffee Shop  
6                 Hotel                   Pub            Yoga Studio  
7    Italian Restaurant         Deli / Bodega    Japanese Restaurant  
8    Italian Restaurant                  Park               Pharmacy  
9           Pizza Place   Distribution Center             Comic Shop

Now we perform clustering on the data for venues by popularity for individual neighborhoods, cluster size selected is 7

In [65]:
# set number of clusters
kclusters = 7

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[:]

array([0, 0, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 5, 0, 1, 0,
       0, 0, 0, 0, 6, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [67]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_rev

# # merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               0           Coffee Shop   
2  43.657162 -79.378937               0        Clothing Store   
3  43.651494 -79.375418               0           Coffee Shop   
4  43.676357 -79.293031               0     Health Food Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Park                Bakery        Breakfast Spot   
1                 Diner      Sushi Restaurant    College Auditorium   
2           Coffee Shop        Cosmetics Shop   Japanese Restaurant   
3                  Café            Restaurant        Cosmetics Shop   
4                 Trail                   Pub         Women's Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0               Theater                  Café                   Pub   
1         Burrito Place                  Café         Smoothie Shop   
2                  Café       Bubble Tea Shop                 Hotel   
3          Cocktail Bar   American Restaurant                   Gym   
4          Dessert Shop    Falafel Restaurant           Event Space   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0           Event Space        Farmers Market                   Shoe Store  
1    Mexican Restaurant              Beer Bar           Italian Restaurant  
2             Bookstore  Fast Food Restaurant    Middle Eastern Restaurant  
3   Moroccan Restaurant      Department Store               Lingerie Store  
4  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant

Resulting Cluster visualization

In [69]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

The final folium visualization shows a dominant cluster 0 out of the seven clusters. 